Sources/SwiftFusion/Inference/ValuesStorage.swift

```diff
@@ -0,0 +1,210 @@
+class AnyVectorStorage: AnyDifferentiableStorage {
+  typealias VectorImplementation = AnyVectorStorageImplementation
+  var vectorImplementation: VectorImplementation {
+    fatalError("implement me!")
+  }
+  
+  /// Adds the vector starting at `otherStart` to `self`.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  final func add(_ otherStart: UnsafeRawPointer) {
+    vectorImplementation.add_(otherStart)
+  }
+  
+  /// Scales each element of `self` by `scalar`.
+  final func scale(by scalar: Double) {
+    vectorImplementation.scale_(by: scalar)
+  }
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  final func dot(_ other: UnsafeRawPointer) -> Double {
+    vectorImplementation.dot_(other)
+  }
+}
+
+/// Contiguous storage of homogeneous `EuclideanVector` values of statically unknown type.
+protocol AnyVectorStorageImplementation:
+  AnyVectorStorage, AnyDifferentiableStorageImplementation
+{
+  /// Adds the vector starting at `otherStart` to `self`.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  func add_(_ otherStart: UnsafeRawPointer)
+  
+  /// Scales each element of `self` by `scalar`.
+  func scale_(by scalar: Double)
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count and element
+  /// type as `self`.
+  func dot_(_ otherStart: UnsafeRawPointer) -> Double
+}
+
+/// APIs that depend on `EuclideanVector` `Element` type.
+extension ArrayStorageImplementation where Element: EuclideanVector {
+  /// Adds the vector starting at `otherStart` to `self`.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count as `self`.
+  func add(_ otherStart: UnsafePointer<Element>) {
+    withUnsafeMutableBufferPointer { elementBuffer in
+      var otherPointer = otherStart
+      for elementIndex in elementBuffer.indices {
+        elementBuffer[elementIndex] += otherPointer.pointee
+        otherPointer = otherPointer.advanced(by: 1)
+      }
+    }
+  }
+  
+  /// Returns the dot product of `self` with the vector starting at `otherStart`.
+  ///
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// Precondition: `otherStart` points to the beginning of a buffer with the same count as `self`.
+  func dot(_ otherStart: UnsafePointer<Element>) -> Double {
+    var result = Double(0)
```

Comment:
Nit: that constructs a `Double` from an `Int`.

```suggestion
    var result: Double = 0
```

or

```suggestion
    var result = 0 as Double
```

Reply:
done